system/lib/wasmfs/file.h

```diff
@@ -150,22 +159,48 @@ class Directory : public File {
   };
 
 private:
-  // Return the file with the given name or null if there is none.
+  // The directory cache, or `dcache`, stores `File` objects for the children of
+  // each directory so that subsequent lookups do not need to query the backend.
+  // It also supports cross-backend mount point children that are stored
+  // exclusively in the cache and not reflected in any backend.
+  enum class DCacheKind { Normal, Mount };
+  struct DCacheEntry {
+    DCacheKind kind;
+    std::shared_ptr<File> file;
+  };
+  // TODO: Use a cache data structure with smaller code size.
+  std::map<std::string, DCacheEntry> dcache;
+
+  // Return the `File` object corresponding to the file with the given name or
+  // null if there is none.
   virtual std::shared_ptr<File> getChild(const std::string& name) = 0;
+
+  // Inserts a file with the given name, kind, and mode. Returns a `File` object
+  // corresponding to the newly created file or nullptr if the new file could
+  // not be created. Assumes a child with this name does not already exist.
+  virtual std::shared_ptr<DataFile> insertDataFile(const std::string& name,
+                                                   mode_t mode) = 0;
+  virtual std::shared_ptr<Directory> insertDirectory(const std::string& name,
+                                                     mode_t mode) = 0;
+  virtual std::shared_ptr<Symlink> insertSymlink(const std::string& name,
+                                                 const std::string& target) = 0;
+
+  // Move the file represented by `file` from its current directory to this
+  // directory with the new `name`, possibly overwriting another file that
+  // already exists with that name. The old directory may be the same as this
+  // directory. On success, return `true`. Otherwise return `false` without
+  // changing any underlying state.
```

Comment:
I think in another thread we discussed that this is only ever called from `rename`. Maybe that's worth mentioning here, plus that we assume the relevant locks have already been taken, as `rename` does?